### ISSUE TITLE
Allow to inject mocked @ConfigProperties classes in quarkus test

### ISSUE THREAD

**Issue by u6f6o** [quarkusio/quarkus#15235]
**Description**

It's currently not possible to mock a config class that was annotated with the `@ConfigProperties` annotation. Let's assume, we have the following setup: 

1.) config properties class
```
@ConfigProperties(prefix = "user-settings")
class UserConfigSnapshot {
    lateinit var minAge: Int
}
```

2.) service that uses the config class
```
@ApplicationScoped
class UserService(private val configSnapshot: UserConfigSnapshot) {

}
```

3.) a test
```
@QuarkusTest
class UserConfigServiceTest {

    private val configSnapshot = mockk<UserConfigSnapshot>()

    @Inject
    lateinit var service: UserConfigService

    @BeforeEach
    fun setupMocks() {
        QuarkusMock.installMockForType(configSnapshot, UserConfigSnapshot::class.java)
    }
}
```
The test fails with the message `UserConfig@2b4a250d is not a normal scoped CDI bean, make sure the bean is a normal scope like @ApplicationScoped or @RequestScoped`. If I annotate the UserConfig with a `@ApplicationScoped` annotation, another error message appears though.

**Comment by radcortez** [comment]
What do you want to mock exactly?

Is there any use case where a mock of `@ConfigProperties` is preferred over setting the configurations you require in the `test` profile? Just trying to understand your use case.

**Comment by u6f6o** [comment]
I wrote a service that performs some computation based on the injected properties and the input parameters provided in the service method. This computation needs to be performed only once per input parameter variation. As I don't need to compute it over and over again, I annotated the service method with the `@CacheResult` annotation and now I wanted to check, if the mock for the config properties is hit only once when I call the service method with the same set of input parameters. 

So it's basically a test, that checks if the cache is working/in place.

**Comment by geoand** [comment]
The reason mocking doesn't work in this case is because the beans created by using `@ConfigProperties` are marked as `Singleton`.

We definitely can make it configurable to allow `@ApplicationScoped` instead of `@Singleton`.
But indeed, what @radcortez makes more sense and should probably be preferred.
What do you think?

**Comment by u6f6o** [comment]
👍  for me it's no show-stopper if we are not able to mock the config properties class.

**Comment by geoand** [comment]
I'll go ahead and close this for now as `won't fix`, but let's keep it in mind if needed for the future

**Comment by maslakov** [comment]
Have the same problem, trying to cover with tests the service logic, which is based on the configuration property with X possible values.
I want to test all X cases, so I would prefer to inititlalize the config value in test. I don't see how to workaround it with config values from properties file - to introduce X profiles ? how should I run different tests in one test class with different profiles?

**Comment by radcortez** [comment]
@maslakov you have a couple of options:

You can use `@TestProfile` like documented here: https://quarkus.io/guides/getting-started-testing#testing_different_profiles

Or you can make a proxy of `@ConfigMapping` like this: https://github.com/ivannov/mocking-config-mappers/pull/1/files. I'll document this piece later.

`@ConfigProperties` was deprecated for 2.0 in favour of `@ConfigMapping`.
